src/Kriti/Parser/Lexer.x

```diff
@@ -0,0 +1,133 @@
+{
+module Kriti.Parser.Lexer where
+
+import qualified Data.Text as T
+import Kriti.Parser.Monad
+import Kriti.Parser.Spans
+import Kriti.Parser.Token
+}
+
+$digit = 0-9
+$alpha = [a-zA-Z]
+$alphanum = [a-zA-Z09]
+
+tokens :-
+
+-- Whitespace insensitive
+<0, expr> $white+                 ;
+
+-- Comments
+<0, expr> "#".*                         ;
+
+-- Syntax
+<0> range                                      { token TokIdentifier }
+<0, expr> true                                 { token (TokBoolLit . (\(Loc sp _) -> Loc sp True)) }
+<0, expr> false                                { token (TokBoolLit . (\(Loc sp _) -> Loc sp False)) }
+<0, expr> \$                                   { token TokIdentifier }
+<0, expr> \$? $alpha [\$ $alpha $digit \_ \-]* { token TokIdentifier }
+
+-- | String Templating
+--
+-- String Templates are represented in our AST as a `StringTem Span
+-- (Vector ValueExt)`. In our lexeme stream we represent this as sequence of tokens:
+--
+-- "www.google.com/{{$path}}" => [TokSymbol SymStringBegin, TokStringLit (Loc _ "www.google.com/"), TokSymbol SymDoubleCurlyOpen _, TokIdentifier (Loc _ "$path"), TokSymbol SymDoubleCurlyClose _, TokSymbol SymStringEnd _]
+--
+-- Everything within the `SymStringBegin` and `SymStringEnd` tokens is
+-- part of the same String Template. Template fragments are wrapped in
+-- `SymDoubleCurlyOpen` and `SymDoubleCurlyClose` tokens.
+--
+-- This representation can be achieved in a single pass of Alex with a
+-- second cleanup pass to concat adjacent `TokStringLit` tokens and is
+-- easy to parse with Happy.
+--
+-- To achieve this we must use two Alex Status Codes: <string> and <expr>.
+--
+-- The core idea is to shift the lexer into <string> when we encounter
+-- '\"' and <expr> when we encounter '{{'. We exit out of <expr> into
+-- <string> when we encounter '}}' and out of <string> into <0> when
+-- we encounter '\"'.
+
+-- Enter String Mode
+<0, expr> \" { \b -> pushStartCode string *> symbol SymStringBegin b}
+
+-- In <string> we have three rules:
+-- 1. Capture a string literal
+<string> (\\ \\ | \\ \` | [^ \" \{ ])+ { token TokStringLit }
```

Comment:
I get really confused when it comes to handling escape characters in strings and am not sure if this regex is totally correct.

Reply:
If you're not sure then it might be worth adding a small `-- NOTE:` comment to that effect, but that's up to you.

Reply:
good call.